sru-exception: add sru exception page for HWE virtualization stack (#631)
ubuntu/ubuntu-project-docs: docs/SRU/reference/exception-Virtualization-Updates.rst

.. _reference-exception-VirtualizationUpdates:

Virtualization Updates
======================

> This document describes the policy for updating virtualization components in a stable,

Reviewer: ```suggestion
This document describes the policy for updating [hardware enablement virtualization components](https://ubuntu.com/server/docs/how-to/virtualisation/virt-hwe/) in a stable,
```

> supported release starting from Ubuntu Resolute 26.04.

Reviewer: ```suggestion
supported release. These opt-in components are available in Ubuntu starting from Ubuntu Resolute 26.04.
```

This cover the following list of source packages:

-  `qemu-hwe <https://launchpad.net/ubuntu/+source/qemu-hwe>`__
-  `libvirt-hwe <https://launchpad.net/ubuntu/+source/libvirt-hwe>`__
-  `edk2-hwe <https://launchpad.net/ubuntu/+source/edk2-hwe>`__
-  `seabios-hwe <https://launchpad.net/ubuntu/+source/seabios-hwe>`__

Rationale
---------

With the advent of hardware-assisted virtualization in modern CPUs, the
virtualization stack (kernel KVM, QEMU and libvirt) evolves quickly alongside
silicon, particularly with the addition of major hardware features such as
confidential computing. Production environments predominantly rely on Ubuntu LTS
releases, whose two-year cadence makes it challenging to deliver the latest
virtualization features to LTS-only users in a timely manner.

To address this, and mirroring the long-standing
:ref:`HWE kernel <https://canonical-kernel-docs.readthedocs-hosted.com/reference/hwe-kernels/>`

> model, Starting from 26.04 LTS, Ubuntu provides a rolling **HWE virtualization stack** for the

Reviewer: ```suggestion
model, Starting from 26.04 LTS, Ubuntu provides an opt-in rolling **HWE virtualization stack** for the
```

Reviewer: You have it below, but it can't be clear enough :-)

LTS that tracks the upstream versions shipped in the latest Ubuntu interim release.

Key notes
---------

- The HWE virtualization stack is **opt-in**

   It is separate from the base stack that remains the default in Ubuntu. Users have to
   explicitly install the HWE packages to switch to the HWE virtualization stack.

- The HWE virtualization stack will be **upgraded once every 6 months**

>    The upgrade follows the same schedule as the HWE kernel. For example, when the next Ubuntu

Reviewer: ```suggestion
   The upgrade follows the same schedule as the [HWE kernel](https://documentation.ubuntu.com/kernel/reference/hwe-kernels/). For example, shortly after the next Ubuntu
```

   interim release (e.g. 26.10) is released, the HWE virtualization stack will be updated in
   the current LTS (e.g. 26.04) to match the versions shipped in that interim release.

> - The HWE virtualization stack rolls to have the **same upstream version the last Ubuntu release**.

Reviewer: ```suggestion
- The HWE virtualization stack rolls to have the **same upstream version that is in the latest Ubuntu release**, it has already been tested and available there for a while to collect additional feedback before rolling its counterpart in the LTS.
```

>    When the HWE stack rolls, it will be upgraded to have the same upstream version as the one shipped in

Reviewer: This bullet point is a duplicate of the others combined - could be removed I guess?

   the latest Ubuntu release. For example, when 26.10 is released, the 26.04 HWE stack will be upgraded
   to have the same upstream version as the one shipped in 26.10.

> - The HWE virtualization stack is **stopped from rolling after 2 years**.

Reviewer: ```suggestion
- The HWE virtualization stack **stops rolling after 2 years** when the next LTS becomes available.
```

   For example, the 26.04 LTS HWE virtualization stack will stop rolling when the next LTS (e.g. 28.04) is released,
   and the last upgrade will be to match the versions shipped in the 28.04 LTS.

Here is the illustrative timeline of a HWE component's upgrade across future Ubuntu releases:

.. image:: https://assets.ubuntu.com/v1/4d67e99d-timeline.png
   :alt: Timeline of the HWE virtualization stack release and upgrade across future Ubuntu releases

Upgrade contents
----------------

- The upgrade will include the **upstream version** along with **relevant bug & security fixes** from the latest Ubuntu release.

> - Some **adjusments** may be needed to the new upstream version to make it fit in the older LTS.

Reviewer: ```suggestion
- Some **adjustments** (e.g. not enabling a particular new feature in the backport) may be needed to the new upstream version to make it build in fit into the older LTS.
```

> - With the above adjustments are not possible or enough to ensure the compability of the virtualization stack, **additional packages** might be needed in the SRU. For example, the build system or other dependencies of the virtualization stack (libvirt-python, virt-manager, meson, ...etc).

Reviewer: ```suggestion
- Sometimes it might happen that adjustments are not enough to ensure the build and compatibility of the virtualization stack. In that case updating it might depend on the availability of **additional packages**, it is expected that build dependencies like toolchains are the most likely to come up like that. We'd need to wait for (by toolchain team) or drive such updates (if the components are on us) to unblock the hardware enablement virtualization stack updates.
```

Risk and mitigation
-------------------

Upgrading the virtualization stack to have a whole new upstream version in a stable release is a major change that
might break our traditional commitments of stability and security in stable releases. This section covers the mitigations
that we put in place to both minimize the risk of such breakage.

1 - The new version will bring in new features, deprecate/modify existing features. These should be detailed both in the file d/NEWS and the SRU bug that is linked to the d/changelog entry.

2 - Any adjustments that are needed to make the new upstream version fit in the older LTS should be detailed in the SRU bug, and should be as minimal as possible to reduce the risk of regressions.

> 3 - The upgrade might break compatiblity or exising usage. This should be avoided by extensive testing:

Reviewer: ```suggestion
3 - The upgrade might break compatibility or existing usage. Where possible that should be avoided and to do so detected by extensive testing:
```

> Since there is complex packaging dependencies between the base and the HWE stacks to ensure they are not mixed, the upgrade

Reviewer: ```suggestion
There are complex packaging dependencies between the *base* and the *HWE* stack, to ensure they are not mixed. The upgrade of the HWE stack could affect the base installation and usage.
```

> might break the base stack installation and usage. Extensive testing should be done to ensure that the upgrade does

Reviewer: ```suggestion
Extensive testing should be done to ensure that the upgrade does
```

> not break the base stack installation and usage, and that switching between the base and HWE stacks with `ubuntu_virt_helper switch`

Reviewer: ```suggestion
not do so, and that switching between the base and HWE stacks with `ubuntu_virt_helper switch`
```

> performs a complete one-to-one replacement, preserving dependencies and each package's auto/manual install mark.

Reviewer: I think we need to tell them "what/how we intend to test" instead of just saying "Extensive testing"

Reviewer: This applies to all of this section including 10 lines up and down I guess

Introducting a new version might affect VM migration as Ubuntu defines new machine types; extensive testing 
on migration is important to ensure we do not break existing migration scenarios, and to identify any necessary
adjustments to the new upstream version to ensure the compatibility of the stack.

The new version might also break runtime compatibility with existing components in the LTS. This runtime
breakage is hard to identify and occurs in production, so we need to do our best to identify and fix it before the
upgrade is released. This is done by the extensive integration test suite that the virtualization components contain,
which covers a wide range of scenarios and use cases, and is ran using the SRU package for each release. The results
of these tests are attached to the SRU bug.

SRU Process
-----------

The SRU should be done with a single bug, for all the source packages. This is because the components

> are tightly coupled and need to be upgraded together to ensure the compatibility of the stack.

Reviewer: ```suggestion
are tightly coupled and should to be upgraded together to ensure the compatibility of the stack.
```

Reviewer: If they are hard depending on being upgraded together they would need to express versioned dependencies - we never needed that for normal upgrades and the same should be here.
Therefore only "should" is it is not as hard to need hard versioned dependencies I'd think

The SRU process should be followed as usual, with the following additional requirements to ensure
that the upgrade is done smoothly and with minimal regressions:

> -  The SRU should be requested per the StableReleaseUpdates

Reviewer: misses Link I guess?

   documented process
-  The template at the end of this document should be used and all
   ‘TODO’ filled out
-  For each release (e.g. resolute, etc.) that is proposed to
   be updated by the SRU a link to the results of integration
   testing, via autopkgtest, successfully completed using the

>    proposed package with no unexplained errors or failures

Reviewer: And I guess here we'd also commit to providing logs of the regression tests (that we have or those you are rewriting)

-  Any packaging changes (e.g. a dependency changes) need to be
   stated

If backwards compatibility is to be broken, this should be clearly

> written at the top of the bug description for the SRU, as well as in the

Reviewer: ```suggestion
written in the bug description for the SRU, as well as in the
```

Reviewer: IMHO "at the top" is a bit too much, but debatable

title with "[breaks-compat]". Furthermore, an email to ubuntu-release

> will be sent to point the release / SRU teams to the bug in order to get

Reviewer: has anyone asked for the "email upfront" ?
It is hard, not all have emails or lists.
Could that not be skipped and just leave the former few lines where you correctly say "his should be clearly written at the top of the bug description for the SRU, as well as in the title with "[breaks-compat]"" ?

approval before uploading to the release's upload queue.

SRU Template
------------

::

   [Impact]

>    This release contains bug fixes, security fixes and/or new features (and, for

Reviewer: I disagree with this
"""
   This release contains bug fixes, security fixes and/or new features (and, for
   the HWE stack, may bring it in line with a newer upstream version) and we
   would like to make sure all of our supported customers have access to these
   improvements. The notable ones are:

   *** <TODO: Create list with LP: # included>

   See the changelog entry below for a full list of changes and bugs.
"""

This template is only for HWE stack updates.
But it reads not like that.
Furthermore we can never decide what is important for whom so we need to also refer to the upstream changelogs for the rest. How about

"""
   This release contains bug fixes, security fixes and new features by bringing
   the hwe virt stack in line with what was released in TODO-XX.YY release. This update
   provides access to those changes by updating the opt-in hwe virtualization stack.

   Notable changes:

   *** <TODO: Create list with LP: # included that would be in but are not in the base release>
   *** <TODO: Pick high level impactful changes from the upstream changelog>
   *** <TODO: Refer to the upstream changelogs that apply between last and new version "For more see ...">
   *** <TODO: This list exists N times, once per component libvirt, qemu, edk2, ...>

>    the HWE stack, may bring it in line with a newer upstream version) and we

Reviewer: Global topic like in the spec document - please rewrite all mentioned of hwe, hwe stack, hwe virt stack, ... to something you pick and then change all occurrences to use the same.

   would like to make sure all of our supported customers have access to these
   improvements. The notable ones are:

   *** <TODO: Create list with LP: # included>

   See the changelog entry below for a full list of changes and bugs.

   [Test Plan]

>    The following development and SRU process was followed:

Reviewer: ```suggestion
   The following additional detail to the SRU process for this case shall hereby be followed:
```

   https://documentation.ubuntu.com/sru/en/latest/reference/exception-Virtualization-Updates

>    Virtualization components contain an extensive integration test suite that is ran using

Reviewer: That autopkgtests need to be good is the default, not worth to mention.
What might be worth to mention is just "The hwe stack packages contain extensive autopkgtests that will run against the builds in -proposed as usual"

   the SRU package for each releases. This test suite's results are available here:
   http://autopkgtest.ubuntu.com/packages/n/virtualization-components
   A successful run is required before the proposed virtualization components package
   can be let into -updates.

>    The virtualization team will be in charge of attaching the artifacts and console

Reviewer: I miss the reference to the regression tests which include the migration and other Ubuntu specific testing - should that be added here too?

   output of the appropriate run to the bug.  Virtualization team members will not

>    mark ‘verification-done’ until this has happened.

Reviewer: "Virtualization team members will not mark ‘verification-done’ until this has happened."

How about reordering this into
"To mark ‘verification-done’ on this update the following has to be completed:"
at the top and then all the others you have as bullet points under that?

   For HWE stack uploads, verification must also confirm that switching between
   the base and HWE variants with `ubuntu_virt_helper switch` performs a complete
   one-to-one replacement, preserving dependencies and each package's auto/manual

>    install mark.

Reviewer: Also mention that we will provide a log of that?

   [Where problems could occur]
   In order to mitigate the regression potential, the results of the
   aforementioned integration tests are attached to this bug.

   The HWE and base stacks are mutually exclusive and the HWE packages are
   separate source packages; particular attention is required to ensure that no
   variant or supported release misses a relevant change, including security
   fixes.

>    <TODO: attach test artifacts for every SRU release, not a link as links expire>

Reviewer: This is part of testing above, not needed here

   [Other Info]

>    <TODO: other background, including which variant (base or HWE) and which

Reviewer: Isn't this always HWE only - this does not fit here for me

Reviewer: Mentioning a reference to the related packages in the latest interim release would be nice.
Like links to their builds that this is based on maybe?

   releases are targeted>

   [Changelog]

>    <TODO: Paste in change log entry, including a link to the upstream release

Reviewer: I suggested to have that above, I'm fine here as well but then move all I suggested above down here please

Upgrade contents
----------------

- The upgrade will include the **upstream version** along with **relevant bug & security fixes** from the latest Ubuntu release.

- Some **adjustments** (e.g. not enabling a particular new feature in the backport) may be needed for the new upstream version to make it build and fit into the older LTS.

- Sometimes it might happen that adjustments are not enough to ensure the build and compatibility of the virtualization stack. In that case updating it might depend on the availability of **additional packages**, it is expected that build dependencies like toolchains are the most likely to come up like that. We'd need to wait for (by toolchain team) or drive such updates (if the components are on us) to unblock the hardware enablement virtualization stack updates.

Risk and mitigation
-------------------

Upgrading the virtualization stack to have a whole new upstream version in a stable release is a major change that
might break our traditional commitments of stability and security in stable releases. This section covers the mitigations
that we put in place to both minimize the risk of such breakage. The mitigations are of 3 kinds:

> - Documentation

Reviewer: those thee could be links to the details below

- Feature drop: case by case decision on dropping support for a particular feature or component if it is too risky to upgrade it.
- Testing

.. _virt-updates-documentation:

Documentation
~~~~~~~~~~~~~
Can you add the link to the Documentation section 
The places where the upgrade is documented are:

  - d/NEWS, d/changelog
  - `Launchpad`_ SRU bug.

>   - `Ubuntu Discourse <UbuntuDiscourse_>`_

Reviewer: no to discourse - you might consider establishing a group of interest instead
https://ubuntu.com/project/docs/SRU/reference/package-specific/#groups-of-sru-interest

For each upgrade, the following information should be documented:

- any adjustments that are needed to make the new upstream version fit in the current LTS.
- any compatibility breakage with existing components or usage.

Features drop
~~~~~~~~~~~~~

> To mitigate the risk of breakage, we might decide to drop support for a particular

Reviewer: ```suggestion
To mitigate the risk of breakage or make it build, we might decide to drop support for a particular
```

feature or component if it is too risky or impossible to have it in the current LTS.
This decision should be made on a case by case basis after careful consideration of the potential impact.

> Testing

Reviewer: So far you describe the intention, what this lacks is calling out how we will do so.
Would it make sense to point to the virtualization regression tests?

I'd expect that your improving rework will land in the same repo, so that pointer would even stay valid.

If you outline things in this section those tests are not yet doing you need to add them.
Until you did add/merge it with the main repo you need to point to the tests helpers you had when developing this.

Example - good: we want to test against migration issues
Example - better: we want to test against migration issues to do so we run this scrip with this config and provide the logfile

~~~~~~~

In addition to the usual testing that the HWE stack naturally inherits

> from the last Ubuntu release (since it has the same upstream version), some specific testing is needed to ensure

Reviewer: ```suggestion
from the last Ubuntu release (since it has the same upstream version), further testing is needed to ensure
```
